File: src/trigram.ts
/**
 * A trigram index, so a content search touches the files that could match
 * instead of every file there is.
 *
 * `grep` and its faster cousins are O(total bytes) per query: they read the
 * whole tree every time you ask. That is fine on a small repo and miserable on
 * a large one, and an agent asks a lot. The alternative is older than any of
 * them — index every overlapping three-byte window once, and at query time
 * intersect the posting lists of the trigrams the pattern must contain. The
 * files that survive are the only ones worth reading. (Design from
 * microsoft/tgrep, which reports up to 52x over ripgrep on very large trees.)
 *
 * The index only ever *narrows*. Every surviving candidate is still matched
 * for real, so a wrong candidate costs time and never correctness.
 *
 * Pure, and no dependencies: a trigram is three bytes packed into a number.
 */

/** `(a << 16) | (b << 8) | c` — injective for three bytes, so no collisions. */
export type Trigram = number;

export function trigramsOf(text: string): Set<Trigram> {
  const out = new Set<Trigram>();
  if (text.length < 3) return out;
  // Latin-1 folding keeps the packing injective for ASCII, which is what
  // source code is; anything above stays distinct enough to narrow with.
  for (let i = 0; i + 2 < text.length; i++) {
    const a = text.charCodeAt(i) & 0xff;
    const b = text.charCodeAt(i + 1) & 0xff;
    const c = text.charCodeAt(i + 2) & 0xff;
    out.add((a << 16) | (b << 8) | c);
  }
  return out;
}

/**
 * What the index can be asked.
 *
 * `all` is the important one: it means the pattern gave nothing indexable, so
 * no candidate set is safe and everything must be read. Getting this wrong is
 * how an index silently starts hiding results.
 */
export type Plan =
  | { kind: "and"; trigrams: Trigram[] }
  | { kind: "or"; branches: Plan[] }
  | { kind: "all" };

/**
 * Literal runs in a regex — the only parts that imply required trigrams.
 *
 * The rule that matters: a run may only contain characters the pattern
 * *requires*, in the order it requires them. Emitting anything else asks the
 * index for trigrams the pattern never promised, and the index then excludes
 * files that genuinely match — a false negative, which is the one failure an
 * index must never produce.
 *
 * That is exactly what an earlier version did. Scanning character by
 * character and merely *breaking* at a delimiter left the delimiter's
 * contents behind as an ordinary run: `x[abcdef]y` yielded `["abcdef"]` and
 * demanded `abc`, `bcd`, `cde`, `def` — none of which a matching `xay`
 * contains. Measured: the file matched under `node`'s own regex and the
 * search returned nothing. `a{2,4}bcdef` had the same shape and required the
 * literal `2,4`.
 *
 * So bracketed and braced spans are skipped over entirely rather than broken
 * at. Losing a run costs narrowing; keeping a wrong one costs the answer.
 *
 * Groups get the same treatment one level up. A quantifier after `)` makes
 * the whole group optional, so nothing pushed since its `(` is required —
 * which needs a stack, because by the time `)?` is seen the group's runs are
 * already in the list. The first fix in this area handled `[...]` and `{...}`
 * and missed exactly this: `(abcd)?xyz` still demanded the trigrams of
 * `abcd`, and a file containing only `xyz` matched the regex while the index
 * excluded it. Lookarounds are dropped the same way, conservatively: a
 * negative `(?!abcd)` must never require what it forbids, and telling the
 * four lookaround forms apart buys too little to be worth the parsing.
 */
export function literalRuns(pattern: string): string[] {
  const runs: string[] = [];
  const groups: Array<{ start: number; discard: boolean }> = [];
  let current = "";
  const breakRun = () => {
    runs.push(current);
    current = "";
  };

  for (let i = 0; i < pattern.length; i++) {
    const ch = pattern[i]!;
    if (ch === "\\") {
      const next = pattern[i + 1];
      // An escaped literal character contributes; a class like \d does not,
      // and ends the run.
      if (next && /[^A-Za-z0-9]/.test(next)) {
        current += next;
        i++;
        continue;
      }
      breakRun();
      i++;
      continue;
    }

    // A character class is an alternation: none of what it holds is required.
    if (ch === "[") {
      breakRun();
      i = skipTo(pattern, i, "]");
      continue;
    }
    if (ch === "(") {
      breakRun();
      let discard = false;
      if (pattern[i + 1] === "?") {
        if (pattern[i + 2] === ":") {
          // (?:…) is an ordinary group in disguise; without consuming the
          // marker here, `?` would read as a quantifier and `:` as text.
          i += 2;
        } else {
          // (?=…) (?!…) (?<=…) (?<!…): zero-width. Only the positive forms
          // truly require their contents, and requiring a negative one's
          // contents excludes exactly the files that match.
          discard = true;
          i += 1;
        }
      }
      groups.push({ start: runs.length, discard });
      continue;
    }
    if (ch === ")") {
      const group = groups.pop();
      const next = pattern[i + 1];
      // `+` is deliberately absent: (abc)+ requires at least one abc.
      if (group && (group.discard || next === "?" || next === "*" || next === "{")) {
        runs.length = group.start;
        current = "";
      } else {
        breakRun();
      }
      continue;
    }
    // A quantifier applies to the character before it, which is therefore not
    // required either — and the count inside the braces is not text to match.
    if (ch === "{") {
      runs.push(current.slice(0, -1));
      current = "";
      i = skipTo(pattern, i, "}");
      continue;
    }
    if (ch === "?" || ch === "*" || ch === "+") {
      runs.push(current.slice(0, -1));
      current = "";
      continue;
    }
    if ("}]|.^$".includes(ch)) {
      breakRun();
      continue;
    }
    current += ch;
  }
  runs.push(current);
  return runs.filter((run) => run.length >= 3);
}

/**
 * Index of the closing delimiter, or the end of the pattern when there is
 * none. An unterminated `[` is a malformed regex the matcher will reject; the
 * planner's job here is only to avoid inventing requirements from it.
 */
function skipTo(pattern: string, from: number, close: string): number {
  for (let i = from + 1; i < pattern.length; i++) {
    if (pattern[i] === "\\") {
      i++;
      continue;
    }
    if (pattern[i] === close) return i;
  }
  return pattern.length;
}

/** A literal pattern: every one of its trigrams must be present. */
export function planForLiteral(literal: string, caseInsensitive: boolean): Plan {
  const text = caseInsensitive ? literal.toLowerCase() : literal;
  const trigrams = [...trigramsOf(text)];
  return trigrams.length === 0 ? { kind: "all" } : { kind: "and", trigrams };
}

/**
 * A regex: take its longest literal run. Using only one run is deliberate —
 * alternation means a run in one branch is not required by the pattern as a
 * whole, and requiring it would hide matches from the other branch.
 */
export function planForRegex(pattern: string, caseInsensitive: boolean): Plan {
  if (pattern.includes("|")) {
    // Each alternative narrows its own branch; the union is safe.
    const branches = pattern.split("|").map((part) => planForRegex(part, caseInsensitive));
    // One unindexable branch can match anywhere, so the union is unbounded.
    if (branches.some((b) => b.kind === "all")) return { kind: "all" };
    return { kind: "or", branches };
  }
  const runs = literalRuns(pattern);
  if (runs.length === 0) return { kind: "all" };
  const longest = runs.reduce((a, b) => (b.length > a.length ? b : a));
  return planForLiteral(longest, caseInsensitive);
}

/** Several patterns: a file matches if any does, so the plans are unioned. */
export function planForPatterns(plans: readonly Plan[]): Plan {
  if (plans.length === 0) return { kind: "all" };
  if (plans.some((p) => p.kind === "all")) return { kind: "all" };
  return plans.length === 1 ? plans[0]! : { kind: "or", branches: [...plans] };
}

/** Trigram → the files containing it. */
export class TrigramIndex {
  private postings = new Map<Trigram, Set<number>>();
  private indexed = new Set<number>();

  add(fileId: number, content: string, caseInsensitive = true): void {
    this.remove(fileId);
    for (const trigram of trigramsOf(caseInsensitive ? content.toLowerCase() : content)) {
      let list = this.postings.get(trigram);
      if (!list) {
        list = new Set();
        this.postings.set(trigram, list);
      }
      list.add(fileId);
    }
    this.indexed.add(fileId);
  }

  remove(fileId: number): void {
    if (!this.indexed.delete(fileId)) return;
    for (const [trigram, list] of this.postings) {
      if (list.delete(fileId) && list.size === 0) this.postings.delete(trigram);
    }
  }

  get size(): number {
    return this.indexed.size;
  }

  has(fileId: number): boolean {
    return this.indexed.has(fileId);
  }

  /**
   * Candidate file ids, or null meaning "no set is safe — read everything".
   * Null and the empty set are different answers and must not be confused:
   * one means everything, the other means nothing.
   */
  candidates(plan: Plan): Set<number> | null {
    if (plan.kind === "all") return null;

    if (plan.kind === "or") {
      const union = new Set<number>();
      for (const branch of plan.branches) {
        const part = this.candidates(branch);
        if (part === null) return null;
        for (const id of part) union.add(id);
      }
      return union;
    }

    // Intersect, smallest posting list first so the working set only shrinks.
    const lists = plan.trigrams.map((t) => this.postings.get(t) ?? new Set<number>());
    if (lists.length === 0) return null;
    lists.sort((a, b) => a.size - b.size);
    let result = new Set(lists[0]!);
    for (const list of lists.slice(1)) {
      if (result.size === 0) break;
      const next = new Set<number>();
      for (const id of result) if (list.has(id)) next.add(id);
      result = next;
    }
    return result;
  }
}

File: native/src/trigram.rs
//! Trigram index over file contents.
//!
//! Every overlapping three-byte window of a file is packed into a `u32` —
//! `(a << 16) | (b << 8) | c` — which is injective, so a trigram is its own
//! hash and no two distinct windows collide. The index maps each trigram to
//! the files containing it, and a search intersects the posting lists of the
//! trigrams its pattern must contain. Only the files that survive are read.
//!
//! The index only ever *narrows*. Every candidate is matched for real
//! afterwards, so a wrong candidate costs time and never correctness — which
//! is what makes it safe to narrow aggressively.

use std::collections::{HashMap, HashSet};

pub type Trigram = u32;

/// A trigram key *is* its own hash, so the default SipHash is pure overhead on
/// a path that runs once per input byte. One multiply-xorshift replaces it;
/// the xorshift is not optional, because hashbrown takes the bucket index from
/// the low bits and the low bits of `value * K` depend only on the last byte.
#[derive(Default, Clone, Copy)]
pub struct TrigramHasher(u64);

impl std::hash::Hasher for TrigramHasher {
    #[inline]
    fn finish(&self) -> u64 {
        self.0
    }
    #[inline]
    fn write(&mut self, bytes: &[u8]) {
        for &b in bytes {
            self.0 = (self.0 ^ u64::from(b)).wrapping_mul(0x0000_0100_0000_01b3);
        }
    }
    #[inline]
    fn write_u32(&mut self, value: u32) {
        let mixed = (u64::from(value)).wrapping_mul(0x9E37_79B9_7F4A_7C15);
        self.0 = mixed ^ (mixed >> 29);
    }
}

#[derive(Default, Clone, Copy)]
pub struct BuildTrigramHasher;

impl std::hash::BuildHasher for BuildTrigramHasher {
    type Hasher = TrigramHasher;
    #[inline]
    fn build_hasher(&self) -> TrigramHasher {
        TrigramHasher::default()
    }
}

type TrigramMap<V> = HashMap<Trigram, V, BuildTrigramHasher>;

/// Every distinct trigram in `bytes`, ASCII-folded so the index and the query
/// agree on case without storing both.
pub fn extract(bytes: &[u8], out: &mut HashSet<Trigram, BuildTrigramHasher>) {
    out.clear();
    if bytes.len() < 3 {
        return;
    }
    for window in bytes.windows(3) {
        let a = window[0].to_ascii_lowercase() as u32;
        let b = window[1].to_ascii_lowercase() as u32;
        let c = window[2].to_ascii_lowercase() as u32;
        out.insert((a << 16) | (b << 8) | c);
    }
}

pub fn of(text: &str) -> Vec<Trigram> {
    let mut set = HashSet::with_hasher(BuildTrigramHasher);
    extract(text.as_bytes(), &mut set);
    let mut list: Vec<Trigram> = set.into_iter().collect();
    list.sort_unstable();
    list
}

/// What the index can be asked. `All` means the pattern gave nothing
/// indexable, so no candidate set is safe and everything must be read —
/// a different answer from "nothing matched", and conflating the two is how
/// an index silently starts hiding results.
#[derive(Debug, Clone)]
pub enum Plan {
    And(Vec<Trigram>),
    Or(Vec<Plan>),
    All,
}

#[derive(Default)]
pub struct Index {
    postings: TrigramMap<Vec<u32>>,
    indexed: HashSet<u32>,
}

impl Index {
    pub fn new() -> Self {
        Self {
            postings: HashMap::with_hasher(BuildTrigramHasher),
            indexed: HashSet::new(),
        }
    }

    pub fn len(&self) -> usize {
        self.indexed.len()
    }

    /// Whether a file's contents are in the index. A listed file that is not is
    /// text kept out of it (too big, empty) and still a grep candidate.
    pub fn contains(&self, file: u32) -> bool {
        self.indexed.contains(&file)
    }

    pub fn add(&mut self, file: u32, trigrams: &[Trigram]) {
        if self.indexed.contains(&file) {
            self.remove(file);
        }
        for &t in trigrams {
            let list = self.postings.entry(t).or_default();
            // Postings stay sorted so intersection is a merge, not a scan.
            //
            // Bulk builds add files in ascending id order, so the new id
            // belongs at the end and one comparison settles it. Falling
            // through to a binary search here would cost fifteen scattered
            // probes on a list held by every file — which measured as most of
            // the time spent building the index.
            match list.last() {
                Some(&last) if last >= file => match list.binary_search(&file) {
                    Ok(_) => {}
                    Err(at) => list.insert(at, file),
                },
                _ => list.push(file),
            }
        }
        self.indexed.insert(file);
    }

    pub fn remove(&mut self, file: u32) {
        if !self.indexed.remove(&file) {
            return;
        }
        self.postings.retain(|_, list| {
            if let Ok(at) = list.binary_search(&file) {
                list.remove(at);
            }
            !list.is_empty()
        });
    }

    /// The index turned back the other way up: every trigram, grouped by the
    /// file it came from, for writing the index out.
    ///
    /// One pass over the postings rather than one pass *per file* — asking each
    /// file separately would rescan the entire index for every file, which on a
    /// large tree is quadratic and would cost more than rebuilding from source.
    pub fn by_file(&self) -> HashMap<u32, Vec<Trigram>> {
        let mut out: HashMap<u32, Vec<Trigram>> = HashMap::with_capacity(self.indexed.len());
        for &file in &self.indexed {
            out.insert(file, Vec::new());
        }
        for (&t, files) in &self.postings {
            for file in files {
                if let Some(list) = out.get_mut(file) {
                    list.push(t);
                }
            }
        }
        for list in out.values_mut() {
            list.sort_unstable();
        }
        out
    }

    /// Candidate file ids, or `None` meaning every file must be read.
    pub fn candidates(&self, plan: &Plan) -> Option<Vec<u32>> {
        match plan {
            Plan::All => None,
            Plan::Or(branches) => {
                let mut union: Vec<u32> = Vec::new();
                for branch in branches {
                    let part = self.candidates(branch)?;
                    union.extend_from_slice(&part);
                }
                union.sort_unstable();
                union.dedup();
                Some(union)
            }
            Plan::And(trigrams) => {
                if trigrams.is_empty() {
                    return None;
                }
                let mut lists: Vec<&Vec<u32>> = Vec::with_capacity(trigrams.len());
                for t in trigrams {
                    match self.postings.get(t) {
                        // A trigram nobody has means nothing can match — an
                        // empty answer, emphatically not "read everything".
                        None => return Some(Vec::new()),
                        Some(list) => lists.push(list),
                    }
                }
                // Smallest first, so the working set only ever shrinks.
                lists.sort_by_key(|l| l.len());
                let mut result = lists[0].clone();
                for list in &lists[1..] {
                    if result.is_empty() {
                        break;
                    }
                    result = intersect(&result, list);
                }
                Some(result)
            }
        }
    }
}

/// Both sides are sorted, so this is a linear merge rather than a lookup loop.
fn intersect(a: &[u32], b: &[u32]) -> Vec<u32> {
    let mut out = Vec::with_capacity(a.len().min(b.len()));
    let (mut i, mut j) = (0usize, 0usize);
    while i < a.len() && j < b.len() {
        match a[i].cmp(&b[j]) {
            std::cmp::Ordering::Equal => {
                out.push(a[i]);
                i += 1;
                j += 1;
            }
            std::cmp::Ordering::Less => i += 1,
            std::cmp::Ordering::Greater => j += 1,
        }
    }
    out
}

/// Literal runs of three or more characters — the only parts of a pattern
/// that imply required trigrams.
///
/// The rule that matters: a run may only contain characters the pattern
/// *requires*, in the order it requires them. Emitting anything else asks the
/// index for trigrams the pattern never promised, and the index then excludes
/// files that genuinely match — a false negative, the one failure an index
/// must never produce.
///
/// Delimited spans (`[…]`, `{…}`) are skipped, not broken at. Groups get the
/// same treatment one level up: a quantifier after `)` makes the whole group
/// optional, so nothing pushed since its `(` is required — which needs a
/// stack, because by the time `)?` is seen the group's runs are already in
/// the list. The first fix here handled classes and braces and missed exactly
/// this: `(abcd)?xyz` still demanded the trigrams of `abcd`, and a file
/// containing only `xyz` matched the regex while the index excluded it.
/// Lookarounds are dropped conservatively — a negative `(?!abcd)` must never
/// require what it forbids. Kept deliberately identical to the TypeScript
/// fallback's `literalRuns`.
pub fn literal_runs(pattern: &str) -> Vec<String> {
    let chars: Vec<char> = pattern.chars().collect();
    let mut runs: Vec<String> = Vec::new();
    let mut groups: Vec<(usize, bool)> = Vec::new();
    let mut current = String::new();
    let mut i = 0usize;
    while i < chars.len() {
        let ch = chars[i];
        if ch == '\\' {
            if let Some(&next) = chars.get(i + 1) {
                if !next.is_ascii_alphanumeric() {
                    current.push(next);
                    i += 2;
                    continue;
                }
            }
            runs.push(std::mem::take(&mut current));
            i += 2;
            continue;
        }
        // A character class is an alternation: nothing it holds is required.
        if ch == '[' {
            runs.push(std::mem::take(&mut current));
            i = skip_to(&chars, i, ']');
            continue;
        }
        if ch == '(' {
            runs.push(std::mem::take(&mut current));
            let mut discard = false;
            if chars.get(i + 1) == Some(&'?') {
                if chars.get(i + 2) == Some(&':') {
                    // (?:…) is an ordinary group in disguise; without
                    // consuming the marker, `?` reads as a quantifier and
                    // `:` as text.
                    i += 2;
                } else {
                    // Lookarounds: only the positive forms truly require
                    // their contents, and requiring a negative one's contents
                    // excludes exactly the files that match.
                    discard = true;
                    i += 1;
                }
            }
            groups.push((runs.len(), discard));
            i += 1;
            continue;
        }
        if ch == ')' {
            let group = groups.pop();
            let next = chars.get(i + 1).copied();
            // `+` is deliberately absent: (abc)+ requires at least one abc.
            let optional = matches!(next, Some('?') | Some('*') | Some('{'));
            match group {
                Some((start, discard)) if discard || optional => {
                    runs.truncate(start);
                    current.clear();
                }
                _ => runs.push(std::mem::take(&mut current)),
            }
            i += 1;
            continue;
        }
        // A quantifier applies to the character before it, which is therefore
        // not required — and the count inside the braces is not text to match.
        if ch == '{' {
            current.pop();
            runs.push(std::mem::take(&mut current));
            i = skip_to(&chars, i, '}');
            continue;
        }
        if matches!(ch, '?' | '*' | '+') {
            current.pop();
            runs.push(std::mem::take(&mut current));
            i += 1;
            continue;
        }
        if matches!(ch, '}' | ']' | '|' | '.' | '^' | '$') {
            runs.push(std::mem::take(&mut current));
            i += 1;
            continue;
        }
        current.push(ch);
        i += 1;
    }
    runs.push(current);
    runs.retain(|r| r.chars().count() >= 3);
    runs
}

/// One past the closing delimiter, or the end of the pattern when there is
/// none. An unterminated `[` is a malformed regex the matcher will reject; the
/// planner's only job is to avoid inventing requirements from it.
fn skip_to(chars: &[char], from: usize, close: char) -> usize {
    let mut i = from + 1;
    while i < chars.len() {
        if chars[i] == '\\' {
            i += 2;
            continue;
        }
        if chars[i] == close {
            return i + 1;
        }
        i += 1;
    }
    chars.len()
}

pub fn plan_for_literal(literal: &str) -> Plan {
    let trigrams = of(literal);
    if trigrams.is_empty() {
        Plan::All
    } else {
        Plan::And(trigrams)
    }
}

pub fn plan_for_regex(pattern: &str) -> Plan {
    if pattern.contains('|') {
        let branches: Vec<Plan> = pattern.split('|').map(plan_for_regex).collect();
        // One branch that can match anywhere makes the whole union unbounded.
        if branches.iter().any(|b| matches!(b, Plan::All)) {
            return Plan::All;
        }
        return Plan::Or(branches);
    }
    let runs = literal_runs(pattern);
    match runs.iter().max_by_key(|r| r.len()) {
        None => Plan::All,
        Some(longest) => plan_for_literal(longest),
    }
}
